src/lib/fixer.ts

```diff
@@ -131,6 +138,164 @@ export class _PmdFixGenerator extends FixGenerator {
+        const diagnosticLine = diagnostic.range.start.line;
+    
+        // Split the text into lines for easier processing
+        const lines = text.split('\n');
+        let classStartLine: number | undefined;
+
+        let inBlockComment = false;
+    
+        // Iterate from the diagnostic line upwards to find the class declaration
+        for (let lineNumber = diagnosticLine; lineNumber >= 0; lineNumber--) {
+            const line = lines[lineNumber];
+
+            // Check if we are in a block comment
+            if (inBlockComment) {
+                if (line.match(this.blockCommentStartPattern)) {
+                    inBlockComment = false;
+                }
+                continue;
+            }
+
+            // Skip single-line comments
+            if (line.match(this.singleLineCommentPattern)) {
+                continue;
+            }
+
+            // Skip block comment in a single line
+            if (line.match(this.blockCommentEndPattern) && line.match(this.blockCommentStartPattern)) {
+                continue;
+            }
+
+            // Check if this line is the start of a block comment
+            if (line.match(this.blockCommentEndPattern)) {
+                inBlockComment = true;
+                continue;
+            }
+            
+            const match = line.match(this.classDeclarationPattern);
+            if (!inBlockComment && match && !this.isWithinQuotes(line, match.index)) {
+                classStartLine = lineNumber;
+                break;
+            }
+        }
+    
+        if (classStartLine !== undefined) {
+            return new vscode.Position(classStartLine, 0);
+        }
+        
+        // Default to the start of the document if class is not found
+        return new vscode.Position(0, 0);
+    }
+
+    /**
+     * Finds the entire line that is one line above a class declaration statement.
+     * @returns The text of the line that is one line above the class declaration.
+     */
+    public findLineBeforeClassStartDeclaration(classStartPosition: vscode.Position, document: vscode.TextDocument): string {
+        // Ensure that there is a line before the class declaration
+        if (classStartPosition.line > 0) {
+            const lineBeforeClassPosition = classStartPosition.line - 1;
+            const lineBeforeClass = document.lineAt(lineBeforeClassPosition);
+            return lineBeforeClass.text;
+        }
+
+        // Return an empty string if it's the first line of the document
+        return '';
+    }
+
+    /**
+     * Helper function to check if match is within quotes
+     * @param line 
+     * @param matchIndex 
+     * @returns 
+     */
+    public isWithinQuotes(line: string, matchIndex: number): boolean {
+        const beforeMatch = line.slice(0, matchIndex);
+        const singleQuotesBefore = (beforeMatch.match(/'/g) || []).length;
+        const doubleQuotesBefore = (beforeMatch.match(/"/g) || []).length;
```

Comment:
Does apex allow double quotes? I thought strings must be single quoted? Not sure if you need to worry about double quotes.

Reply:
Strings have to be single-quoted. I genuinely don't remember if you're allowed to use double-quotes as string boundaries.
BUT you might want to check whether the string containing escaped single-quotes throws off the algorithm. E.g., `'this is a string. it\'s got a single-quote inside it'`.